### ISSUE TITLE
refactor: add implementation

### ISSUE THREAD

**Issue by Trass3r** [clangd/clangd#445]
When clicking on a function declaration it should offer add implementation and add implementation in corresponding .cpp file.

**Comment by sam-mccall** [comment]
Yep. We have the "define-outline" code action which is very closely related, so a workaround is:
 - replace `;` with `{}`
 - trigger "define outline"

But we should support this directly. @kadircet there's no reason this would be harder, right?

**Comment by sam-mccall** [comment]
Oh, there is a catch: it's kind of bad to offer this if a definition exists.

This is mostly relevant in a header file, where the existing definition (if it exists) isn't visible through the AST.
"define-outline" triggers on definitions, so we can just assume there's no *other* definition and offer the refactoring.
However "add implementation" triggers on declarations, and we need to find out if the definition already exists.

We can't necessarily afford to query the index to see if a definition already exists. Or rather, currently refactorings are independent, and having them all query the index to check whether to *offer* the refactoring doesn't scale. Maybe we need to add some mechanism for refactorings to coordinate index requests, or maybe we should just have the "framework" look up the decl under the cursor in all cases.

**Comment by kadircet** [comment]
Yes there were 2 concerns; first is the one you mentioned, cost of querying the index in all of the tweaks during prepare. we were thinking about coming up with a "short-lived cached-index" that could be shared among all prepares (for a single code-action invocation) as in theory most of the checks should be querying the same symbols, but this was a bold claim at the time (and I think still is).

the second one is, even if we query index we might still be missing the definition, e.g. bg-index is off or stale. and then code action will result in broken code (hopefully a compile error, but it might even introduce ODR violations).

**Comment by Trass3r** [comment]
Does VSCode/LSP support a two-stage approach? First some generic entry like 'Add/Move Implementation...' and only then it queries and shows the options in a new selection menu?

**Comment by kadircet** [comment]
clangd currently offers all of its refactoring actions through a mechanism called "code-actions" which is indicated by a lightbulb icon in vscode whenever available. So as @sam-mccall suggested you can try something like:

in a.h:
```
struct A {
  void foo() { return; }
};
```
and also have a file called a.cpp (action is normally more clever than that) and put your cursor over `foo` in the header. Then you should see a lightbulb with subselections containing "move function definition out-of-line".

**Comment by Trass3r** [comment]
Didn't get any define outline code action with clangd 10, neither in the .h nor when putting the class in the .cpp.

Anyway, I also just had the need for the opposite refactoring, move out-of-line implementation into the class body.

**Comment by kadircet** [comment]
> Didn't get any define outline code action with clangd 10, neither in the .h nor when putting the class in the .cpp.

yeah looks like we forgot to flip the flag of define out-of-line in clangd-10 release :( sorry for the confusion.
you would need a top of the tree build, or use one of our weekly builds at https://github.com/clangd/clangd/releases
or you could try passing `--hidden-features` to clangd, note that this will result in having some actions always available (for debugging purposes hence the lightbulb will be always with you).

> Anyway, I also just had the need for the opposite refactoring, move out-of-line implementation into the class body.

we also have that one, but it is quite conservative. it will refuse to move any function if a symbol within the body "might" not be visible at the declaration location.

**Comment by Trass3r** [comment]
Will test it sooner or later. Hopefully it also works with template classes.

**Comment by ammen99** [comment]
> yeah looks like we forgot to flip the flag of define out-of-line in clangd-10 release :( sorry for the confusion.
> you would need a top of the tree build, or use one of our weekly builds at https://github.com/clangd/clangd/releases
> or you could try passing `--hidden-features` to clangd, note that this will result in having some actions always available (for debugging purposes hence the lightbulb will be always with you).
> 
> >

Hi, currently running the latest pre-release build (clangd_snapshot_20201018), and I am not seeing this code action with coc-nvim. Is there any other flag I need to set for newer clangd versions?

**Comment by kadircet** [comment]
nope, can you provide logs? preferably on a separate bug and while passing `--log=verbose` to clangd.

**Comment by ammen99** [comment]
> nope, can you provide logs? preferably on a separate bug and while passing `--log=verbose` to clangd.

Alright, I opened https://github.com/clangd/clangd/issues/568 with my testcase.

**Comment by tom-anders** [comment]
Would it maybe be an option to add a protocol extension for this? Something like textDocument/implementMethod you'd trigger with the cursor on a function definition. We'd only need to query the index once the client sends the request and possibly return an error if there's already a definition somewhere.

**Comment by KKoovalsky** [comment]
@kadircet After two years, are the concerns still valid? I.e. is querying the index in `prepare()` still expensive, or it got cheaper during the progression of clangd? Was some kind of the aforementioned "short-lived cached-index" introduced? What I want to ask really: can we progress with this feature?

Regarding the ODR violations, maybe we could rely in that case on the compiler/linker to properly resolve them, so that the risk that we generate another copy of the function definition could be mitigated? With the `DefineOutline` tweak, we assume that no definition exists without query, thus there are certain assumptions we can make and "live with them"?

I have implemented a feature like this [as a standalone command line tool](https://github.com/KKoovalsky/Tsepepe#function-definition-generator) on top of `libclang`, and later introduced [a VIM plugin which utilizes it](https://github.com/KKoovalsky/TsepepeVim#function-definition-generation). Yet, I feel that it could easily be a `CodeAction` that could be used by and LSP client working with clangd.

**Comment by ckandeler** [comment]
I'd like to revive this discussion, as I believe this is valuable functionality that users rightfully expect to exist.
As to the two concerns mentioned:
  a) The performance argument seems to be of the "slippery slope" kind. Of course you don't want every tweak to query
      the index, but as of now, there would be exactly one of them, and it would have a very good reason to do so.
      I'd argue that it suffices to keep an eye on this for new tweaks, and if a second or third one comes along that
      can claim in good conscience that it needs to query the index, we can still look for a general solution. (And because
      we will have several uses cases by then, we will have a better idea about the requirements, i.e. there's less danger
      of introducing an infrastructure that misses the mark).
  b) I'm not convinced by the argument that we might break the code if the index is out of date. That can happen with any
      action that changes code across translation units, most prominently symbol renaming.

**Comment by tom-anders** [comment]
One note from my side regarding broken code resulting from this:

(Sam)
> This is mostly relevant in a header file, where the existing definition (if it exists) isn't visible through the AST.
> "define-outline" triggers on definitions, so we can just assume there's no _other_ definition and offer the refactoring.
> However "add implementation" triggers on declarations, and we need to find out if the definition already exists.

(kadircet)
> the second one is, even if we query index we might still be missing the definition, e.g. bg-index is off or stale. and then code action will result in broken code (hopefully a compile error, but it might even introduce ODR violations).

At least in my experience, 99% percent of the time when I want to apply this tweak, _I know that no definitions exists yet_, since I just added the declaration to the header file, and now want to generate the definition in the `.cpp`.  Are there any other workflows I'm missing here? Will people really try to trigger this on function declarations which already have a definition?

**Comment by ckandeler** [comment]
> At least in my experience, 99% percent of the time when I want to apply this tweak, _I know that no definitions exists yet_, since I just added the declaration to the header file, and now want to generate the definition in the `.cpp`. Are there any other workflows I'm missing here? Will people really try to trigger this on function declarations which already have a definition?

At the very least, it would spam the list of available tweaks and potentially add bogus light bulbs, which would certainly be annoying. However, with the index querying it should be a rare occurrence.

**Comment by tom-anders** [comment]
> > At least in my experience, 99% percent of the time when I want to apply this tweak, _I know that no definitions exists yet_, since I just added the declaration to the header file, and now want to generate the definition in the `.cpp`. Are there any other workflows I'm missing here? Will people really try to trigger this on function declarations which already have a definition?
> 
> At the very least, it would spam the list of available tweaks and potentially add bogus light bulbs, which would certainly be annoying. However, with the index querying it should be a rare occurrence.

Oh right I forgot about the light bulb, I do not have that configured in my editor.

**Comment by arashrun** [comment]
In vscode, There was a extention call [c-mantic](https://marketplace.visualstudio.com/items?itemName=tdennis4496.cmantic) has implemented the feature.
In nvim, Has a plugin [nvim-treesitter-cpp-tools](https://github.com/Badhi/nvim-treesitter-cpp-tools)
In vim, I used the `treesitter` + `coc.nvim` wrote a plugin that implements the simplest feature of adding a definition, where the user can decide whether to copy the definition to the clipboard or to a file, the plug-in is still in the development stage there should be a lot of cases that have not been taken into account, just to provide an idea of implementation  [coc-mantic](https://github.com/arashrun/coc-mantic)

**Comment by tbertels** [comment]
Could this feature also do the reverse? Add a function declaration (in the right place) when clicking on a function definition.

**Comment by HighCommander4** [comment]
> Could this feature also do the reverse? Add a function declaration (in the right place) when clicking on a function definition.

Are you thinking of a member or non-member function here?

The member function case is trickier since the starting point would be invalid code if a declaration does not yet exist.

**Comment by vimpostor** [comment]
> The member function case is trickier since the starting point would be invalid code if a declaration does not yet exist.

Is it really as simple as that, why can't it be solved by a fixit quick action on the error itself?

For example if I use a function from an unimported header, clangd can still produce a quickfix action to add that include, despite it clearly being invalid code.

**Comment by cprn** [comment]
Unpopular opinion - name the code-action `add implementation (experimental)` and risk breaking the user's code. Since querying an index is non-trivial, a trivial check can be made as follows:

1. if in a header file (e.g. `foo.h`)
2. look for corresponding code file (e.g. `foo.cpp`)
3. if the code file doesn't exist, doesn't include the header or doesn't contain the definition
4. offer a code-action that adds the definition in that code file
5. maybe even add header include in that code file
6. maybe even create that code file (i.e. `foo.cpp` for `foo.h`)

This check should be quick, light bulbs should be displayed correctly in the majority of cases, and no-one should be surprised if the code-action triggers an ODR violation. Obviously, it'd be a temporary measure until another code-action requiring the "short-lived cached-index" emerges.

**Comment by C3js189** [comment]
> It works for me now. Though it would be good if VSCode jumped to the position in the .cpp where the function got inserted. Hopefully it will also work with template classes.
> 
> Related functionality is [Create Method Implementations](https://docs.wholetomato.com/default.asp?W458), [Implement interface](https://docs.wholetomato.com/default.asp?W165) and maybe something that helps with PIMPLing a class.





https://stackdev.space/questions/defining-a-member-function-outside-of-the-class-definition

**Comment by jclsn** [comment]
So is this implemented partially or not at all? I don't have any other code actions that "fix error under cursor" from clangd.  I am using coc.nvim. Was actually hoping for a bit more.

**Comment by HighCommander4** [comment]
> So is this implemented partially or not at all?

The closest thing currently available is what's discussed in [this comment](https://github.com/clangd/clangd/issues/445#issuecomment-655786531).

>  I don't have any other code actions that "fix error under cursor" from clangd.

(I'm a bit confused by this because the code action discussed in this issue is not one that generally fixes an error, except perhaps in the presence of `constexpr` which requires a definition to be present in the current translation unit.)

**Comment by jclsn** [comment]
I guess coc.nvim doesn't support this define-outline code action then. At least nothing is triggered, when I exchange ``;`` with ``{}`` or does the mean the LSP command with this?

**Comment by HighCommander4** [comment]
I'm not a neovim user (and in general, questions about how to trigger an LSP feature in a particular client are better discussed in the issue tracker of that client) but some searching around led me to [this](https://superuser.com/questions/1737168/neovim-keybinding-for-lsp-code-action-with-options):

> The default keybind for code actions is `<space>ca`, after pressing which a menu pops up.

**Comment by HighCommander4** [comment]
One thing to note is the code action's availability depends on the cursor position. Based on experimentation in VSCode, the cursor needs to be over the method's name for the "define outline" code action to be available.

**Comment by jclsn** [comment]
I tried several vim LSP clients now and it doesn't work. The code action also isn't advertised, as you can see from the [yegappan/lsp](https://github.com/yegappan/lsp) output here. Is it possible that this code action needs a custom request to be sent to the server? If yes, can someone show me the documentation where to find this? I can't find `define-outline` here https://clangd.llvm.org/extensions

```
'clangd' Language Server Capabilities
=====================================
astProvider: true
callHierarchyProvider: true
clangdInlayHintsProvider: true
codeActionProvider: {'codeActionKinds': ['quickfix', 'refactor', 'info']}
compilationDatabase: {'automaticReload': true}
completionProvider: {'resolveProvider': false, 'triggerCharacters': ['.', '<', '>', ':', '"', '/', '*']}
declarationProvider: true
definitionProvider: true
documentFormattingProvider: true
documentHighlightProvider: true
documentLinkProvider: {'resolveProvider': false}
documentOnTypeFormattingProvider: {'moreTriggerCharacter': [], 'firstTriggerCharacter': '
'}
documentRangeFormattingProvider: true
documentSymbolProvider: true
executeCommandProvider: {'commands': ['clangd.applyFix', 'clangd.applyRename', 'clangd.applyTweak']}
foldingRangeProvider: true
hoverProvider: true
implementationProvider: true
inactiveRegionsProvider: true
inlayHintProvider: true
memoryUsageProvider: true
referencesProvider: true
renameProvider: true
selectionRangeProvider: true
semanticTokensProvider: {'full': {'delta': true}, 'legend': {'tokenTypes': ['variable', 'variable', 'parameter', 'function', 'method', 'function', 'property', 'variable', 'class', 'interface', 'enum', 'enumMember', 'type', 'type', 'unknown', 'namespace', 'typeParameter', 'concept', 'type', 'macro', 'modifier', 'operator', 'bracket', 'label', 'comment'], 'tokenModifiers': ['declaration', 'definition', 'deprecated', 'deduced', 'readonly', 'static', 'abstract', 'virtual', 'dependentName', 'defaultLibrary', 'usedAsMutableReference', 'usedAsMutablePointer', 'constructorOrDestructor', 'userDefined', 'functionScope', 'classScope', 'fileScope', 'globalScope']}, 'range': false}
signatureHelpProvider: {'triggerCharacters': ['(', ')', '{', '}', '<', '>', ',']}
standardTypeHierarchyProvider: true
textDocumentSync: {'save': true, 'change': 2, 'openClose': true}
typeDefinitionProvider: true
typeHierarchyProvider: true
workspaceSymbolProvider: true
~additionalInitResult_offsetEncoding: 'utf-32'
~additionalInitResult_serverInfo: {'version': 'clangd version 20.1.8 linux x86_64-pc-linux-gnu', 'name': 'clangd'}
```

**Comment by HighCommander4** [comment]
> The code action also isn't advertised, as you can see from the [yegappan/lsp](https://github.com/yegappan/lsp) output here.

I'm not sure what you're expecting to see in that output that you're not.

The "define outline" code action is covered by:

```
codeActionProvider: {'codeActionKinds': ['quickfix', 'refactor', 'info']}
```

Note that the `codeActionKinds` are categories of code actions, not individual code actions, and "define outline" falls into the category `refactor`. Code actions are not advertised individually.

> Is it possible that this code action needs a custom request to be sent to the server?

No, it uses the standard `textDocument/codeAction`.

> I tried several vim LSP clients now and it doesn't work. 

For further diagnosis, please share the following:

 * a complete code example
 * verbose [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs) from a session where you tried to invoke the code action
 * indicate what position in the source file you tried to invoke the code action at

(And since this issue is about adding a new code action that's (slightly) different from "define outline", please share the above in a new issue rather than this one.)

**Comment by jclsn** [comment]
@HighCommander4 Nevermind, it was my bad. I did not read that I needed to exchange `;` with `{}`. Then it works. Sorry for wasting your time.

Coming back to this issue. I would also expect an `add-implementation` code action to be offered when having `;`  to then add an empty implementation to the source file.

**Comment by timon-ul** [comment]
I have decided to merge the associated PR now, to address the original concerns:

1. Scalability:
Yes the current approach does not scale, but as long as this is the only code action that does an index lookup in the `prepare` stage, there is imo no point in investing ressources into creating some sort of cache or condensed lookup, this will be addressed when (or even if at all) we get a 2nd code action that requires this.

2. Stale index:
Yes there are scenarios where the index can go stale, but then on top the developer has to be unaware of this in the mean time added Definition and I think this is highly unlikely. The tradeoff between never being able to create any incorrect code state and not getting a useful feature seems to be for me much more in favour of getting said feature. Also maybe this is a difference in mindset, but in general I do not expect a code action to always create something correct, there are other parts of `clangd`where I think an incorrect result is way more problematic and they all suffer when we have a stale index. In general clangd's usefulness decreases with a stale index.
